File: aisongtool/zimage.py
"""Z-Image-Turbo (https://huggingface.co/Tongyi-MAI/Z-Image-Turbo) — optional
background-image generation for the Create flow's "generate from the song's
prompt" option, installed into its own isolated `uv` env (see tools_install.py
ENV_SPECS["zimage-uv"]) the same way demucs-uv/whisperx-uv are.

Unlike ACE-Step (a persistent API server with its own multi-minute model-load
timing problems), each Z-Image generation is a one-shot subprocess that loads
the model, makes one image, saves it, and exits — no server lifecycle to
manage, just a command for `jobs.run_blocking` to run to completion.
"""
from __future__ import annotations

from pathlib import Path

from .paths import data_dir, workers_dir
from .toolrunner import venv_python


class ZImageError(RuntimeError):
    pass


def dest_dir() -> Path:
    return data_dir() / "zimage-uv"


def is_synced() -> bool:
    return venv_python(dest_dir()).exists()


def install(log=print) -> None:
    """Provision the isolated zimage-uv env (`tools_install.ENV_SPECS`) —
    same one-call shape as `ace_step.install()`, kept here so callers don't
    need to know it's just another `tools_install.ensure_env` entry."""
    from .tools_install import default_gpu_mode, ensure_env

    ensure_env("zimage-uv", default_gpu_mode(), log=log)


# Backgrounds sit behind centered lyric text, so a busy/photorealistic image
# fights for attention with the subtitles — a consistent minimalist sky/cloud
# aesthetic (think 7CLOUD album-cover style art) keeps every generated
# background calm and legible regardless of the song's own prompt. See
# desktop/src/main/tools/zimage.ts for the same suffix.
_STYLE_SUFFIX = (
    ", minimalistic red sky, soft gradient clouds, dreamy pastel sky background, "
    "7cloud album cover aesthetic"
)


def build_generate_cmd(
    prompt: str, out_path: Path, width: int = 1280, height: int = 720,
    seed: int | None = None,
) -> list[str]:
    if not is_synced():
        raise ZImageError(
            "Z-Image-Turbo isn't installed yet. Run `aisongtool install-tool z-image` "
            "(or use the Setup tab) first."
        )
    py = venv_python(dest_dir())
    cmd = [
        str(py), str(workers_dir() / "zimage_generate.py"),
        "--prompt", prompt + _STYLE_SUFFIX, "--out", str(out_path),
        "--width", str(width), "--height", str(height),
    ]
    if seed is not None:
        cmd += ["--seed", str(seed)]
    return cmd

File: desktop/src/main/tools/zimage.ts
/** Port of `aisongtool/zimage.py` — Z-Image-Turbo, a one-shot subprocess
 * (no server lifecycle like ACE-Step), already installed via the Setup
 * view's "Install Z-Image Turbo" button. */
import { existsSync } from 'fs'
import path from 'path'
import { envDir, venvPython, workersDir } from '../paths'

export class ZImageError extends Error {}

export function destDir(): string {
  return envDir('zimage-uv')
}

export function isSynced(): boolean {
  return existsSync(venvPython(destDir()))
}

// Backgrounds sit behind centered lyric text, so a busy/photorealistic
// image fights for attention with the subtitles — a consistent minimalist
// sky/cloud aesthetic (think 7CLOUD album-cover style art) keeps every
// generated background calm and legible regardless of the song's own prompt.
const STYLE_SUFFIX =
  ', minimalistic red sky, soft gradient clouds, dreamy pastel sky background, 7cloud album cover aesthetic'

function buildGenerateCmdRaw(
  prompt: string,
  outPath: string,
  width: number,
  height: number,
  seed: number | null
): string[] {
  if (!isSynced()) {
    throw new ZImageError(
      "Z-Image-Turbo isn't installed yet. Install it from the Setup view first."
    )
  }
  const cmd = [
    venvPython(destDir()),
    path.join(workersDir(), 'zimage_generate.py'),
    '--prompt',
    prompt,
    '--out',
    outPath,
    '--width',
    String(width),
    '--height',
    String(height)
  ]
  if (seed !== null) cmd.push('--seed', String(seed))
  return cmd
}

export function buildGenerateCmd(
  prompt: string,
  outPath: string,
  width = 1280,
  height = 720,
  seed: number | null = null
): string[] {
  return buildGenerateCmdRaw(prompt + STYLE_SUFFIX, outPath, width, height, seed)
}

/** Same generation, but the prompt is used exactly as given — no style
 * suffix appended. Used by the "Minimalistic Sky" template's hardcoded
 * "Minimalistic red sky" prompt, which is already the whole point. */
export function buildGenerateCmdLiteral(
  prompt: string,
  outPath: string,
  width = 1280,
  height = 720,
  seed: number | null = null
): string[] {
  return buildGenerateCmdRaw(prompt, outPath, width, height, seed)
}

/** `zimage_gradio.py` — a standalone Gradio UI that loads the model once
 * and serves repeated generations, the same shape as ACE-Step's own
 * `acestep` entry point. */
export function buildGuiCmd(port = 7861): string[] {
  if (!isSynced()) {
    throw new ZImageError("Z-Image-Turbo isn't installed yet. Install it from the Setup view first.")
  }
  return [venvPython(destDir()), path.join(workersDir(), 'zimage_gradio.py'), '--port', String(port)]
}
